Add an iterator over XFB rows

linkmauve opened the pull request:
This emits slices of `u16`, and allows users to not use unsafe for drawing.

This is based on #21, since it wouldn’t build on current nightly otherwise.

linkmauve committed:
Add an iterator over XFB rows

This emits slices of u16, and allows users to not use unsafe for
drawing.

ProfElements commented on luma_core/src/vi.rs:
```diff
@@ -5,20 +5,19 @@
 /// sent to the screen, in YUYV format.  It must be allocated as contiguous physical memory.
 pub struct Xfb {
-    data: Pin<Box<[u8]>>,
+    data: Box<[u16]>,
```
The data needs to be pinned for the duration of the use of the XFB due to mmio usage. Rust doesn't confirm that the pointer wont move.

linkmauve replied:
Oh right, thanks, I’ve fixed that issue.

ProfElements reviewed (changes requested):
Besides one small issue it all looks awesome!

linkmauve committed:
Replace removed from_exposed_addr_mut() with without_provenance_mut()

Now that the `exposed_provenance` and `strict_provenance` features have
been stabilized, we can migrate to its definitive functions.

This change is undefined behaviour though, but `as *mut u32` was similarly
UB, so we just trade one UB for another…

`core::ptr`’s documentation says:
> Situations where a valid pointer must be created from just an address,
> such as baremetal code accessing a memory-mapped interface at a fixed
> address, are an open question on how to support. These situations will
> still be allowed, but we might require some kind of “I know what I’m
> doing” annotation to explain the situation to the compiler. It’s also
> possible they need no special attention at all, because they’re
> generally accessing memory outside the scope of “the abstract machine”,
> or already using “I know what I’m doing” annotations like “volatile”.

ProfElements commented on luma_core/src/vi.rs:
```diff
@@ -37,23 +37,66 @@ impl Xfb {
+impl<'a> Iterator for XfbIterMut<'a> {
+    type Item = &'a mut [u16];
+
+    fn next(&mut self) -> Option<&'a mut [u16]> {
```
would `self.xfb.data.chunks_mut(self.xfb.stride_in_u16()).next()` work here?

linkmauve replied:
I’ve ended up replacing the whole `XfbIterMut` struct with `ChunksMut`, which already does the exact same and more.

ProfElements commented:
I think there is a better way to implement `iter_mut` either with `impl Iterator<Item = &'a mut [u16]>` or with `core::iter::from_fn`

ProfElements reviewed (approved):
Looks good :)

ProfElements committed:
Merge branch 'main' into safer-vi
